kernel/src/vm/vmar/mod.rs

```diff
@@ -690,6 +705,114 @@ impl Vmar_ {
         &self.vm_space
     }
 
+    fn read_maybe_remote(&self, vaddr: Vaddr, writer: &mut VmWriter) -> Result<usize> {
+        let len = writer.avail();
+        let read = |frame: UFrame, skip_offset: usize| {
+            let mut reader = frame.reader();
+            reader.skip(skip_offset);
+            let read_bytes = reader.read_fallible(writer)?;
+
+            Ok(read_bytes)
+        };
+
+        self.access_maybe_remote(vaddr, len, PageFlags::R, read)
+    }
+
+    fn write_maybe_remote(&self, vaddr: Vaddr, reader: &mut VmReader) -> Result<usize> {
+        let len = reader.remain();
+        let write = |frame: UFrame, skip_offset: usize| {
+            let mut writer = frame.writer();
+            writer.skip(skip_offset);
+            let written_bytes = writer.write_fallible(reader)?;
+
+            Ok(written_bytes)
+        };
+
+        self.access_maybe_remote(vaddr, len, PageFlags::W, write)
+    }
+
+    /// Access memory at `vaddr..vaddr+len` within the VMAR using `op`.
+    ///
+    /// The `VmSpace` corresponding to the VMAR is not required be activated on
+    /// the current CPU. If any page in the range is not mapped or does not have
+    /// the required page flags, a page fault will be handled to try to make the
+    /// page accessible.
+    fn access_maybe_remote<F>(
+        &self,
+        vaddr: Vaddr,
+        len: usize,
+        required_page_flags: PageFlags,
+        mut op: F,
+    ) -> Result<usize>
+    where
+        F: FnMut(UFrame, usize) -> Result<usize>,
+    {
+        let range = {
+            let Some(end) = vaddr.checked_add(len) else {
+                return_errno_with_message!(Errno::EINVAL, "address overflow");
+            };
+            if !is_userspace_vaddr(vaddr) || !is_userspace_vaddr(end - 1) {
+                return_errno_with_message!(Errno::EINVAL, "invalid user space address");
+            }
+            vaddr.align_down(PAGE_SIZE)..end.align_up(PAGE_SIZE)
+        };
+
+        let preempt_guard = disable_preempt();
+        let vmspace = self.vm_space();
+        let mut cursor = vmspace.cursor(&preempt_guard, &range)?;
+        let mut current_va = range.start;
+        let mut bytes = 0;
+
+        while current_va < range.end {
+            cursor.jump(current_va)?;
+            let (_, mut item) = cursor.query()?;
+
+            if item
+                .as_ref()
+                .is_none_or(|item| !item.prop().flags.contains(required_page_flags))
+            {
+                drop(cursor);
```

Comment:
Need to `drop(preempt_guard)`. `handle_page_fault` cannot be called with preemption disabled.